File: src/utils/lru.ts
/**
 * Simple LRU cache. Nothing fancy, just enough for tracking
 * content hashes across requests without leaking memory.
 */
export class LRU<K, V> {
  private map = new Map<K, V>()
  private maxSize: number

  constructor(maxSize: number = 100) {
    this.maxSize = maxSize
  }

  get(key: K): V | undefined {
    const value = this.map.get(key)
    if (value === undefined) return undefined

    // move to end (most recent)
    this.map.delete(key)
    this.map.set(key, value)
    return value
  }

  set(key: K, value: V): void {
    if (this.map.has(key)) {
      this.map.delete(key)
    } else if (this.map.size >= this.maxSize) {
      // evict oldest
      const oldest = this.map.keys().next().value!
      this.map.delete(oldest)
    }
    this.map.set(key, value)
  }

  has(key: K): boolean {
    return this.map.has(key)
  }

  get size(): number {
    return this.map.size
  }

  clear(): void {
    this.map.clear()
  }
}

File: python/cachellm/utils/lru.py
from collections import OrderedDict
from typing import Generic, Optional, TypeVar

K = TypeVar("K")
V = TypeVar("V")


class LRU(Generic[K, V]):
    """Simple LRU cache. Nothing fancy, just enough for tracking
    content hashes across requests without leaking memory."""

    def __init__(self, max_size: int = 100):
        self._dict: OrderedDict[K, V] = OrderedDict()
        self._max_size = max_size

    def get(self, key: K) -> Optional[V]:
        if key not in self._dict:
            return None
        self._dict.move_to_end(key)
        return self._dict[key]

    def set(self, key: K, value: V) -> None:
        if key in self._dict:
            del self._dict[key]
        elif len(self._dict) >= self._max_size:
            self._dict.popitem(last=False)
        self._dict[key] = value

    def has(self, key: K) -> bool:
        return key in self._dict

    @property
    def size(self) -> int:
        return len(self._dict)

    def clear(self) -> None:
        self._dict.clear()
